File: sdk/python/substrait_compliance/engine.py
"""Engine interface and metadata classes."""

from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from typing import Dict, List, Optional
from .table_data import TableData
from .result import ComplianceResult


@dataclass
class EngineInfo:
    """Metadata about a Substrait engine."""
    
    name: str
    version: str
    vendor: str
    description: str = ""
    
    def __str__(self) -> str:
        return f"{self.name} {self.version} by {self.vendor}"


@dataclass
class EngineCapabilities:
    """Describes what Substrait features an engine supports."""
    
    supported_relations: List[str] = field(default_factory=list)
    supported_functions: List[str] = field(default_factory=list)
    supported_types: List[str] = field(default_factory=list)
    max_plan_depth: int = 100
    supports_extensions: bool = True
    
    def supports_relation(self, relation_type: str) -> bool:
        """Check if engine supports a relation type."""
        return relation_type in self.supported_relations
    
    def supports_function(self, function_name: str) -> bool:
        """Check if engine supports a function."""
        return function_name in self.supported_functions


class ComplianceEngine(ABC):
    """
    Interface that engines must implement for compliance testing.
    
    Engines implement this interface to participate in decentralized
    compliance testing.
    """
    
    @abstractmethod
    def get_info(self) -> EngineInfo:
        """Return engine metadata."""
        pass
    
    @abstractmethod
    def get_capabilities(self) -> EngineCapabilities:
        """Return engine capabilities."""
        pass
    
    @abstractmethod
    def execute_plan(
        self,
        plan_bytes: bytes,
        input_data: Dict[str, TableData]
    ) -> ComplianceResult:
        """
        Execute a Substrait plan with given input data.
        
        Args:
            plan_bytes: Binary Substrait plan
            input_data: Map of table name to table data
            
        Returns:
            ComplianceResult with execution outcome
        """
        pass
    
    @abstractmethod
    def validate_plan(self, plan_bytes: bytes) -> ComplianceResult:
        """
        Validate a Substrait plan without executing it.
        
        Args:
            plan_bytes: Binary Substrait plan
            
        Returns:
            ComplianceResult with validation outcome
        """
        pass

File: sdk/rust/src/engine.rs
//! Engine interface and metadata types.

use crate::error::Result;
use crate::result::ComplianceResult;
use crate::table_data::TableData;
use std::collections::HashMap;

/// Metadata about a Substrait engine.
#[derive(Debug, Clone)]
pub struct EngineInfo {
    pub name: String,
    pub version: String,
    pub vendor: String,
    pub description: String,
}

impl EngineInfo {
    pub fn new(name: impl Into<String>, version: impl Into<String>, vendor: impl Into<String>) -> Self {
        Self {
            name: name.into(),
            version: version.into(),
            vendor: vendor.into(),
            description: String::new(),
        }
    }
    
    pub fn with_description(mut self, description: impl Into<String>) -> Self {
        self.description = description.into();
        self
    }
}

impl std::fmt::Display for EngineInfo {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{} {} by {}", self.name, self.version, self.vendor)
    }
}

/// Describes what Substrait features an engine supports.
#[derive(Debug, Clone, Default)]
pub struct EngineCapabilities {
    pub supported_relations: Vec<String>,
    pub supported_functions: Vec<String>,
    pub supported_types: Vec<String>,
    pub max_plan_depth: usize,
    pub supports_extensions: bool,
}

impl EngineCapabilities {
    pub fn new() -> Self {
        Self {
            max_plan_depth: 100,
            supports_extensions: true,
            ..Default::default()
        }
    }
    
    pub fn supports_relation(&self, relation_type: &str) -> bool {
        self.supported_relations.iter().any(|r| r == relation_type)
    }
    
    pub fn supports_function(&self, function_name: &str) -> bool {
        self.supported_functions.iter().any(|f| f == function_name)
    }
}

/// Main trait that engines must implement for compliance testing.
pub trait ComplianceEngine {
    /// Return engine metadata.
    fn get_info(&self) -> EngineInfo;
    
    /// Return engine capabilities.
    fn get_capabilities(&self) -> EngineCapabilities;
    
    /// Execute a Substrait plan with given input data.
    fn execute_plan(
        &self,
        plan_bytes: &[u8],
        input_data: &HashMap<String, TableData>,
    ) -> Result<ComplianceResult>;
    
    /// Validate a Substrait plan without executing it.
    fn validate_plan(&self, plan_bytes: &[u8]) -> Result<ComplianceResult>;
}
